torchao/prototype/autoround/core.py

```diff
@@ -0,0 +1,231 @@
+    sym: bool = False
+    iters: int = 200
+    group_size: int = 128
+    train_bs: int = 4
+    eval_bs: int = 4
+    seed: int = 42
+    amp: bool = True
+    nsamples: int = 128
+    seqlen: int = 2048
+    quant_lm_head: bool = False
+
+
+auto_round_config = AutoRoundConfig()
+
+
+@torch.no_grad()
+def create_qmodel_from_qdq_model(qdq_model: torch.nn.Module):
+    """Create a quantized model from a qdq model.
+
+    The qdq_model includes Linear quantized by auto-round, which includes qdq weight, scale, zp.
+    """
+
+    @torch.no_grad()
+    def apply_static_quant(observed_linear):
+        device = observed_linear.weight.device
+        scale = observed_linear.scale.to(device)
+        zero_point = observed_linear.zp.to(device)
+
+        def to_uintx_weight(input_float):
+            quant_min = 0
+            quant_max = auto_round_config.bits**2 - 1
+            block_size = (1, observed_linear.group_size)
+            from torchao.dtypes.uintx.Uintx import UintxLayoutType
+            from torchao.quantization.quant_primitives import (
+                MappingType,
+                ZeroPointDomain,
+            )
+
+            pack_dim = -1
+            bit_width = auto_round_config.bits
+            layout_type = UintxLayoutType(bit_width=bit_width, pack_dim=pack_dim)
+            return to_affine_quantized_static(
+                input_float=input_float,
+                scale=scale.to(torch.bfloat16),
+                zero_point=zero_point,
+                block_size=block_size,
+                target_dtype=torch.uint8,
+                quant_min=quant_min,
+                quant_max=quant_max,
+                zero_point_domain=ZeroPointDomain.INT,
+                layout_type=layout_type,
+            )
+
+        def to_int4_tinygemm_weight(input_float):
+            # TODO(Yi): check the weight shape, `group_size`, and `inner_k_tiles` to make sure the tinygemm can handle it
+            inner_k_tiles = 8
+            quant_min = 0
+            quant_max = auto_round_config.bits**2 - 1
+            # Shift the zeros to align with tiny gemm.
+            # The dequantization process in tiny gemm:
+            #   tiny_dequant = (tiny_quant - 8) * scale + tiny_zp
+            # The dequantization porcess in auto-round
+            #   dequant = (quant - zp) * scale
+            # To align with tiny gemm:
+            #   dequant = (quant - 8 + 8 - zp) * scale
+            #           = (quant - 8) * scale + (8 - zp) * scale
+            #              \__/                 \______________/
+            #            tiny_quant                 tiny_zp
+            mid_point = (quant_max + quant_min + 1) / 2
+            shifted_zero_point = (mid_point - zero_point) * scale
+            block_size = (1, observed_linear.group_size)
+            orig_out_features, orig_in_features = input_float.shape
+            in_features = find_multiple(orig_in_features, 1024)
+            out_features = find_multiple(orig_out_features, 8)
```

Comment:
would it be easier if padding is applied before quantization? this is what we did for tinygemm: https://github.com/pytorch/ao/blob/2c8e3f368613f896fde14cc24b9268c9e125575e/torchao/dtypes/affine_quantized_tensor.py#L358

Reply:
I think we need to keep padding after quantization. If we pad `weight` before quantization(optimization), it may cause the padding values to non zero, as we would also update `zero` and `scale` for the padding part.
If I understand correctly, AO's `woq_int4` won't cause such issues, as it uses RTN to determine the `zero` and `scale`, while the `auto-round` uses the gradient.